File: src/main/java/io/github/nikpivkin/lifetime/LifetimeInternalListener.java
package io.github.nikpivkin.lifetime;

import io.github.nikpivkin.lifetime.events.GodModeChangedEvent;
import io.github.nikpivkin.lifetime.events.LifetimeChangedEvent;
import io.github.nikpivkin.lifetime.events.PluginStateChangedEvent;
import org.bukkit.Bukkit;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;

public class LifetimeInternalListener implements Listener {

  private final LifetimeTaskManager taskManager;

  public LifetimeInternalListener(LifetimeTaskManager taskManager) {
    this.taskManager = taskManager;
  }

  @EventHandler
  public void onGodModeEnabled(GodModeChangedEvent.Enabled event) {
    taskManager.cancelTask(event.getPlayer());
  }

  @EventHandler
  public void onGodModeDisabled(GodModeChangedEvent.Disabled event) {
    taskManager.startupTask(event.getPlayer());
  }

  @EventHandler
  public void onPluginEnabled(PluginStateChangedEvent.Enabled event) {
    taskManager.startupTask(Bukkit.getOnlinePlayers());
  }

  @EventHandler
  public void onPluginDisabled(PluginStateChangedEvent.Disabled event) {
    taskManager.cancelTask(Bukkit.getOnlinePlayers());
  }

  @EventHandler
  public void onLifetimeChanged(LifetimeChangedEvent event) {
    if (event.getNewValue().isOut()) {
      taskManager.cancelTask(event.getPlayer()); // TODO kick user ?
    }
  }
}

File: src/main/kotlin/io/github/nikpivkin/lifetime/LifetimeInternalListener.kt
package io.github.nikpivkin.lifetime

import io.github.nikpivkin.lifetime.events.GodModeChangedEvent
import io.github.nikpivkin.lifetime.events.LifetimeChangedEvent
import io.github.nikpivkin.lifetime.events.PluginStateChangedEvent
import org.bukkit.Bukkit
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener

class LifetimeInternalListener(
    private val lifetimeTaskManager: LifetimeTaskManager
) : Listener {

    @EventHandler
    fun onGodModeEnabled(event: GodModeChangedEvent.Enabled) {
        lifetimeTaskManager.cancelTask(event.player)
    }

    @EventHandler
    fun onGodModeDisabled(event: GodModeChangedEvent.Disabled) {
        lifetimeTaskManager.startupTask(event.player)
    }

    @EventHandler
    fun onPluginEnabled(event: PluginStateChangedEvent.Enabled) {
        lifetimeTaskManager.startupTask(Bukkit.getOnlinePlayers())
    }

    @EventHandler
    fun onPluginDisabled(event: PluginStateChangedEvent.Disabled) {
        lifetimeTaskManager.cancelTask(Bukkit.getOnlinePlayers())
    }

    @EventHandler
    fun onLifetimeChanged(event: LifetimeChangedEvent) {
        if (event.newValue.isOut) {
            lifetimeTaskManager.cancelTask(event.player) // TODO kick user ?
        }
    }
}
